src/pulling_gauge.rs

```diff
@@ -0,0 +1,93 @@
+use std::{collections::HashMap, fmt, sync::Arc};
+
+use crate::{
+    core::Collector,
+    proto::{Gauge, Metric, MetricFamily, MetricType},
+};
+use protobuf::RepeatedField;
+
+/// A [Gauge] that returns the value from a provided function on every collect run.
+///
+/// This metric is the equivalant of Go's
+/// <https://pkg.go.dev/github.com/prometheus/client_golang@v1.11.0/prometheus#GaugeFunc>
+///
+/// # Examples
+/// ```
+/// # use prometheus::{Registry, PullingGauge};
+/// let registry = Registry::new();
+/// let gauge = PullingGauge::new("available_parallelism", "The available parallelism, usually the numbers of logical cores.", Box::new(|| std::thread::available_parallelism().unwrap().get() as f64)).unwrap();
```

Comment:
Thanks for adding a doc-example here, it's very useful. Perhaps it could be a bit more readable by splitting this line into multiple one and pulling out the polling function beforehand.

Unfortunately it looks like `std::thread::available_parallelism()` is more recent than our current MSRV. We could bump it here, or we could try to rework it. What about using `std::sync::Once::is_completed()` instead and showing it before and after completion?

Reply:
Nice trick!